File: knowledge-mobile/sync/aggregator.ts
import * as Crypto from 'expo-crypto';

import { getSyncConceptRows } from '../db/queries';

export type SyncPayload = {
  course_id: string;
  week: string;
  concepts: Array<Record<string, unknown>>;
  payload_hash: string;
};

export async function buildSyncPayload(courseId: string, week: string): Promise<SyncPayload> {
  const rows = await getSyncConceptRows();
  const concepts = rows.map((r) => ({
    concept: r.concept,
    status: r.status,
    clearing_method: r.clearing_method,
    lens_signature: r.lens_signature,
    integrity_suspect: Boolean(r.integrity_suspect),
  }));

  const payload = { course_id: courseId, week, concepts };
  const canonical = JSON.stringify(payload);
  const payload_hash = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, canonical);
  return { ...payload, payload_hash };
}

File: knowledge/sync/aggregator.py
import hashlib
import json
from datetime import UTC, datetime
from typing import Dict, Any, List

from .. import config
from .. import db


def current_week_label(now: datetime | None = None) -> str:
    dt = now or datetime.now(UTC)
    year, week, _ = dt.isocalendar()
    return f"{year}-W{week:02d}"


def _concept_record(row: Dict[str, Any]) -> Dict[str, Any]:
    # Privacy guard: do not include source_text, chat history, or typing metrics.
    return {
        "concept": row.get("concept", ""),
        "status": row.get("status", "on_loan"),
        "clearing_method": row.get("clearing_method") or "",
        "lens_signature": row.get("lens_signature") or "",
        "integrity_suspect": bool(row.get("integrity_suspect", 0)),
    }


def build_weekly_payload() -> Dict[str, Any]:
    rows = db.list_sync_concepts()
    concepts: List[Dict[str, Any]] = [_concept_record(r) for r in rows]
    payload = {
        "course_id": config.COURSE_ID,
        "week": current_week_label(),
        "concepts": concepts,
    }
    canonical = json.dumps(payload, separators=(",", ":"), sort_keys=True)
    payload["payload_hash"] = hashlib.sha256(canonical.encode("utf-8")).hexdigest()
    return payload
